bdai_ros2_wrappers/bdai_ros2_wrappers/subscription.py

```diff
@@ -0,0 +1,73 @@
+# Copyright (c) 2023 Boston Dynamics AI Institute Inc.  All rights reserved.
+import threading
+import typing
+
+import rclpy.node
+import rclpy.qos
+import rclpy.task
+
+import bdai_ros2_wrappers.process as process
+
+MessageT = typing.TypeVar("MessageT")
+
+
+def wait_for_message_async(
+    msg_type: MessageT,
+    topic_name: str,
+    *,
+    qos_profile: typing.Union[rclpy.qos.QoSProfile, int] = 1,
+    node: typing.Optional[rclpy.node.Node] = None,
+) -> rclpy.task.Future:
+    """
+    Wait for message on a given topic asynchronously.
+
+    Args:
+        msg_type: type of message to wait for.
+        topic_name: name of the topic to wait on.
+        qos_profile: optional QoS profile for temporary topic subscription.
+        node: optional node for temporary topic subscription, defaults to
+        the current process-wide node (if any).
+
+    Returns:
+        A future for the incoming message.
+
+    Raises:
+        RuntimeError: if no node is available.
+    """
+    node = node or process.node()
+    if node is None:
+        raise ValueError("No process-wide ROS 2 node available (did you use bdai_ros2_wrapper.process.main?)")
+    future = rclpy.task.Future()
+
+    def callback(msg: MessageT) -> None:
+        if not future.done():
+            future.set_result(msg)
+
+    sub = node.create_subscription(msg_type, topic_name, callback, qos_profile)
```

Comment:
Should there be something that removes the subscription when this is done?

Reply:
There is! Right below, when the future is done.

Reply:
Ah, skimmed too quickly

Reply:
Ah, skimmed too quickly